department-of-veterans-affairs/caseflow
# Jonathan/APPEALS-4533: Route VSO Users Back to Case Details Page on Submit/Cancel Hearing Conversion

Resolves [APPEALS-4533](https://vajira.max.gov/browse/APPEALS-4533)

### Description
On the Hearings Details page (`hearings/components/Details.jsx`) the cancel and submit button onClick methods are modified for all VSO users in the following fashions:
- `Cancel Button`: This will navigate a VSO user back to the Case Details page for the appeal the hearing is associated with.
- `Submit/"Convert Hearing to Virtual" Button`: This will also navigate a VSO to the Case Details page, and a success banner will appear on the page as visual confirmation that the hearing in question has been converted to virtual.

### Acceptance Criteria
`On the Hearings Convert to Virtual Form`:
- [ ] When a VSO user selects the cancel option, user is redirected to the Case Details page.
- [ ] When a non-VSO user selects the cancel option with the hearing conversion form open, they are returned to the Hearing Details page (these pages are at the same url path; there is some conditional rendering that takes place in the React component).
- [ ] When a VSO user clicks the "Convert Hearing to Virtual" button, they are redirected to the Case Details page.
- [ ] Whenever a VSO user is redirected to the Case Details page after converting a hearing to virtual, a success banner appears at the top of the page.
- [ ] When a non-VSO user clicks the "Convert Hearing to Virtual" button on the hearing conversion form, they are returned to the Hearing Details page.

### Testing Plan
1. Open the Rails console, and grab a non-virtual hearing that has a disposition of `nil`:
```ruby
h = Hearing.where(disposition: nil).reject{|h| h.virtual? || h.scheduled_for_past?}.first
```

2. Get the UUID of the appeal the hearing is associated with:
```ruby
h.appeal.uuid
```

3. Log into Caseflow as a VSO user.

4. Navigate to the Case Details for the appeal to load this page on top of the web history stack:
`/queue/appeals/<appeal-uuid>`

5. Navigate to the "Convert Hearing to Virtual" form in Caseflow Hearings for the hearing from step 1:
```ruby
h.uuid
```

`/hearings/<hearing-uuid>/details`

6. You should now be on the "Convert Hearing to Virtual" form. Scroll down to the bottom of the page, and click on the `Cancel` link in the bottom-left corner.

7. You should be back at the Case Details page from step 4.

8. Go back to the "Convert Hearing to Virtual" form:
`/hearings/<hearing-uuid>/details`

9. Fill in all of the required form fields (Veteran/Appellant Email, Veteran/Appellant Timezone, Representative Email, and Representative Timezone).

10. Click on the "Convert to Virtual Hearing" button in the bottom-right corner of the page.

11. You will once again be at the Case Details page. You should see the following success banner (with a different appellant name):
<img width="834" alt="success-conversion-to-virtual" src="https://user-images.githubusercontent.com/99351305/173140392-52b7b69c-6215-45db-8591-b43df6ae8aa3.PNG">

12. Make sure that the hearing is now virtual:
```ruby
h.reload
h.virtual?
```

13. Next, switch to a hearings coordinator user (ex: `BVASYELLOW`).

14. Obtain a UUID for new non-virtual hearing with a `nil` disposition:

```ruby
Hearing.where(disposition: nil).reject(&:virtual?).first.uuid
```

15. Navigate to the Hearings Details page:
`/hearings/<hearing-uuid>/details`

16. You should not have been immediately brought to the "Convert Hearing to Virtual" form as you were as a VSO user. You will need to select "Virtual" from the "Hearing Type" dropdown in order to bring up the form.

17. Once you have arrived at the conversion form, scroll down to the bottom of the page and click on the "Cancel" link in the bottom-left corner. You should be brought back to the Hearings Details page, and not the Case Details page as you were as a VSO user.

18. Select "Virtual" from the "Hearing Type" dropdown once more. Enter values into all of the fields, and click on the "Convert to Virtual Hearing" button in the bottom-right of the page. You should be brought back to the Hearings Details page, and a success banner should be visible.


### User Facing Changes

 On Cancel|On Submission
 ---|---
![vso-cancel-conversion](https://user-images.githubusercontent.com/99351305/173134374-5f914791-c867-4cae-bc6e-7d06e5352e2e.gif)|![vso-complete-conversion](https://user-images.githubusercontent.com/99351305/173134399-c0c99442-e2b1-4ed0-9fb5-085b0966871c.gif)


### Code Documentation Updates
